Remove right negative margin from pinned items

t-hamano opened the pull request:
## What?

This PR removes the negative margin on the right from pinned items in the header. This should ensure that all icons have the same gap of 8px.

## Why?

This style was added in #40411 to fix gaps, but now it seems unnecessary.

## Testing Instructions

Try a scenario like the one below.

- Site Editor, Post Editor, Widget Editor
- "Show button text labels" enabled/disabled
- Mobile layout
- Custom Pinned Item

To test Custom Pinner Item, run the code below in your browser console.

Post Editor:

```javascript
wp.plugins.registerPlugin( 'my-plugin', {
	render: function () {
		return wp.element.createElement(
			wp.editPost.PluginSidebar,
			{
				name: 'my-plugin',
				icon: 'admin-post',
				title: 'My plugin',
			},
			'Meta field'
		);
	},
} );
```

Site Editor:

```javascript
wp.plugins.registerPlugin( 'my-plugin', {
	render: function () {
		return wp.element.createElement(
			wp.editSite.PluginSidebar,
			{
				name: 'my-plugin',
				icon: 'admin-post',
				title: 'My plugin',
			},
			'Meta field'
		);
	},
} );
```

## Screenshots or screencast <!-- if applicable -->

| State | Before | After |
|--------|--------|--------|
| Default | ![before](https://github.com/WordPress/gutenberg/assets/54422211/f2d2d1b6-a9a4-4f25-9e68-0b61a9cae2f1) | ![after](https://github.com/WordPress/gutenberg/assets/54422211/69d526b0-68f3-441a-988d-f18972af1f58) |
| Focused | ![before-focused](https://github.com/WordPress/gutenberg/assets/54422211/1808476c-bfce-4052-b68d-43e35ac382b9) | ![after-focused](https://github.com/WordPress/gutenberg/assets/54422211/49951752-037c-402f-9b1c-3b3dac631131) |
| Show Text | ![before-text](https://github.com/WordPress/gutenberg/assets/54422211/177e485c-f5b5-4d9c-ade8-a3672cfecd6c) | ![after-text](https://github.com/WordPress/gutenberg/assets/54422211/b71a8eac-4ab2-408e-a500-ca73f0f7ded4) |
| Show Text (Focused) | ![before-text-focused](https://github.com/WordPress/gutenberg/assets/54422211/795e1394-eb76-4892-8cda-f3707f6045a5) | ![after-text-focused](https://github.com/WordPress/gutenberg/assets/54422211/a02835ad-e3c0-4110-9e09-9ed82ebb2824) |
| Custom Pinned Item | ![before-plugin-sidebar](https://github.com/WordPress/gutenberg/assets/54422211/7791515a-9054-4d28-95e7-b2ea0653710a) | ![after-plugin-sidebar](https://github.com/WordPress/gutenberg/assets/54422211/e196fb0c-0f68-45ab-bdeb-2ce491685396) |

t-hamano committed:
Remove right negative margin from pinned items

jameskoster commented:
@jasmussen will know for sure but I think this style may have been added to visually balance the position of the icons accounting for the minimal Options (ellipsis) icon. 

For me the inconsistent spacing between these buttons is equally awkward, so I would approve of this change because it simplifies things a little. But let's see what Joen says.

jasmussen commented:
Yes that's right. The vertical ellipsis due to the thinness of the icon is better optically balanced with the negative margin. We used the negative margin instead of a smaller width, to keep the touchable footprint of the ellipsis the same size. I wonder if there's a better way to document this aspect of the vertical ellipsis button.

jameskoster commented:
I'm not sure how sustainable it is to apply that convention holistically. For instance in the data views we have actions at the end of each table row, with the ellipsis serving as an 'overflow menu':

<img width="360" alt="Screenshot 2024-01-09 at 10 36 02" src="https://github.com/WordPress/gutenberg/assets/846565/fa53b4c7-206d-4f38-aaa6-58d7c0a1194c">

If we applied the negative margin here then the action icons lose alignment with the pagination buttons. 

There's a similar story in the Site Editor when the Styles panel is open. Note the slight mis-alignment of icons:

<img width="279" alt="Screenshot 2024-01-09 at 10 38 05" src="https://github.com/WordPress/gutenberg/assets/846565/41ebc273-c2c3-48f1-865b-47f665994766">

I'm not sure this trade-off is really worth it given the maintenance burder. I'd be inclined to opt for a more consistent convention, e.g. make all 'toolbar' icon buttons 32px, with 8px gap:

<img width="271" alt="Spacing" src="https://github.com/WordPress/gutenberg/assets/846565/697bb064-3001-4603-88b5-677a4d34e158">

Then see if we can balance the ellipsis another way. Perhaps try the horizontal orientation, or dare I say a _slightly_ heavier stroke?

---

Alternatively the negative margin should probably be built into the component somehow, so that it doesn't need to be applied manually across all icon buttons that use the vertical ellipsis.

jasmussen commented:
I do agree there's potential for simplification here. 

What if the ellipsis button was always 24x32? Or what if the ellipsis button had a `margin-left: -4px; margin-right: -4px;`? I don't think we should change the shape or design of the icon.

t-hamano commented:
@jasmussen @jameskoster Thank you for your wonderful review.

>What if the ellipsis button was always 24x32? Or what if the ellipsis button had a `margin-left: -4px; margin-right: -4px;`?

It will look something like this: Narrowed the button width from 32px to 24px and removed the negative margin on the left element.

Furthermore, since the Option button appears to be closer to the right side, I added 4px to the padding on the right side.

Personally, I find this style a little cramped 😅

![image](https://github.com/WordPress/gutenberg/assets/54422211/82d5f68a-0c77-4d66-8ba4-c07843c0da6f)
![image](https://github.com/WordPress/gutenberg/assets/54422211/3af3fe41-0b0a-423b-945f-fdc6a94df6fc)

jasmussen commented:
> Personally, I find this style a little cramped 😅

Feels cramped to me only when focused, whereas the resting state which you'd see most of the time feels off balance. Would the negative left/right margins on the ellipsis have any side effects? Or does it have to be a negative left margin on the ellipsis, since presumably the ellipsis is always last in a series?

jameskoster commented:
The only method I can think of is to add a style to the button component that applies the negative margin to the last button in a series when the icon is an ellipsis. But that's not bullet-proof because an ellipsis could appear next to a group of buttons in a container, in which case the negative margin wouldn't be applied.

It seems very tricky to write a style that will account for all situations, but if there's a smarter way to build this behavior into the component then it would be good to explore. Otherwise it doesn't feel worth the hassle to me. There will always be circumstances where the negative margin isn't applied which creates inconsistency.

t-hamano commented:
Personally, I'm hesitant to put negative margin on the `Button` component itself. Because I think that whether the margin is positive or negative, it intervenes in the context outside of the component.

The main reason I submitted this PR is that when I enable "Show button text labels", the spaces between the buttons look very cramped.

![image](https://github.com/WordPress/gutenberg/assets/54422211/b47b799e-e6a6-45de-bd2a-9876ca6455ce)

For example, how about an approach that overrides the current negative margin to zero only when "Show button text labels" is enabled?

jasmussen commented:
It seems like there are some strong preferences for being generic and not having exceptions, I think @richtabor also stumbled on this, so I'm happy to defer to you all on this.

t-hamano commented:
@jasmussen @jameskoster Thank you for your deep insight. I would like to merge this PR, but there may be an opportunity to reconsider this negative margin in the future.
